src/client-joo/reactive_html5.ml

```diff
@@ -0,0 +1,327 @@
+      (* buttons must be in a group for justification to work *)
+      let a =
+        (Option.value_map ~default:[] ~f:(fun c ->
+             if enabled then [a_onclick c] else []) on_click)
+        @ [
+          a_class
+            ((if enabled then [] else ["disabled"])
+             @ ["btn"; "btn-default"; ]);
+        ] in
+      button ~a content |> in_group
+
+
+    let pagination items =
+      nav [
+        ul ~a:[a_class ["pagination"]]
+          (List.map items ~f:(function
+             | `Disabled content ->
+               disabled_li content
+             | `Enabled (on_click, content) ->
+               li [a ~a:[a_onclick on_click] content]
+             ))
+      ]
+
+    let panel ~body =
+      div ~a:[ a_class ["not-container-fluid"]] [
+        div ~a:[ a_class ["panel"; "panel-default"]] [
+          div ~a:[ a_class ["panel-body"; ]] body
+        ];
+      ]
+
+    let table_responsive ~head ~body =
+      div ~a:[a_class ["table-responsive"]] [
+        tablex
+          ~thead:head
+          ~a:[a_class ["table"; "table-condensed";
+                       "table-bordered"; "table-hover"]] [
+          tbody body
+        ]
+      ]
+
+    let collapsable_ul
+        ?(ul_kind = `Inline) ?(maximum_items = 4) items =
+      let make_ul_content items =
+        List.map items ~f:(fun s -> li [s]) in
+      let list_style =
+        match ul_kind with
+        | `None -> []
+        | `Inline -> ["list-inline"; "inline-items-separated"]
+      in
+      match List.length items with
+      | n when n <= maximum_items ->
+        ul ~a:[a_class list_style] (make_ul_content items)
+      | n ->
+        let expanded = Reactive.Source.create false in
+        Reactive_node.ul
+          ~a:[a_class list_style]
+          Reactive.(
+            Source.signal expanded
+            |> Signal.map ~f:(fun expandedness ->
+                let button =
+                  a ~a:[
+                    a_onclick (fun _ ->
+                        Reactive.Source.set expanded (not expandedness);
+                        false);
+                  ] [
+                    pcdata (if expandedness then "⊖" else "⊕")
+                  ] in
+                match expandedness with
+                | true -> (make_ul_content (items @ [button]))
+                | false ->
+                  let shown_items = List.take items maximum_items @ [button] in
+                  (make_ul_content shown_items)
+              )
+            |> Signal.list
+          )
+
+
+    let collapsable_pre ?(first_line_limit = 30) string =
+      match String.find string ~f:((=) '\n') with
+      | None -> (None, pre [pcdata string])
+      | Some end_of_first_line ->
+        let expanded = Reactive.Source.create false in
+        let content_signal =
+          Reactive.Source.signal expanded
+          |> Reactive.Signal.map ~f:(function
+            | true -> string
+            | false ->
+              (String.sub_exn string ~index:0
```

Comment:
just a minor remark, you might achieve the same with css
http://www.w3schools.com/cssref/css3_pr_text-overflow.asp